### ISSUE TITLE
Implement focus management for chooser modals

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#5338]
Originally reported in #4199. Related: #5336, #5337

It should be possible for screen reader users to interact with chooser dialogs for images, documents, snippets, embeds, links.

Context from #4199:

> Blind people rely entirely on the keyboard to access their system, so the mouse is out of the question. They use Tap and the arrow keys to find their way around, while the browser respond to these navigation commands by setting the keyboard focus on each element according to its tabindex

> If you want to embed an image on a page, a file browser dialog opens, again, the user is unaware of its appearance because the focus is still. Once the focus is inside the modal, it should be trapped there. There should also be a way to move between tabs within the modal.

At the moment, here are specific issues users encounter:

- When opening a chooser dialog, there is no indication that the dialog has opened
- Focus doesn't move to the chooser dialog once opened
- Focus isn't trapped in the chooser modals once there

**Comment by thibaudcolas** [comment]
Katie from Torchbox has had a look at some potential improvements we could make to those modals to make them more accessible & usable. I’ll copy her recommendations over here for future reference. Some might go beyond strict "focus management" but they feel relevant nonetheless.

## Image & document upload modal behaviour

Issues:

 - The user is not aware of the modal as the focus is still when opening.
 - The voice over announcement is not as helpful as what it could be.
 - ESC is not closing the modal

Recommendations:

 - Ensure that the focus is recognising there is a modal shown on the UI
 - Give context on what’s going on, i.e “You are now on ‘choose an image’ modal” before listing actions within that modal. 

![Screenshot 2019-06-24 at 16 59 54](https://user-images.githubusercontent.com/877585/60034025-9d08a000-96a1-11e9-9f57-9e31ae837310.png)

## Image & document upload modal tab order

Recommended tab order:

- Once the user has learned from the screen reader what the modal is about, they should have the -option to close the modal within the first few tabs rather having to go all the way to the bottom to close the modal.
- We should make sure that ESC on the keyboard closes the modal. 

![Screenshot 2019-06-24 at 17 02 43](https://user-images.githubusercontent.com/877585/60034157-dd681e00-96a1-11e9-9b6b-3456e42b1f9f.png)

## Image & document upload modal: other identified issues 

Issues:

 - The screen reader does not give help text context that’s related to the field
 - The screen reader doesn’t say there is a ‘Choose file’ button and says there is invalid data would could confuse the user

Recommendations:

 - If possible, outline the additional text associated with the field so that the user is aware of the required information associated with the field 
 - Reflect that there is a ‘Upload’ button action within voiceover screen reader

![Screenshot 2019-06-24 at 17 03 28](https://user-images.githubusercontent.com/877585/60034202-fb358300-96a1-11e9-8ea6-4bdc176164a0.png)

## Image & document upload modal: considerations

- "Upload images" – Consider renaming the field name to something more meaningful. This would not only potentially help screen readers but also avoid potential confusion for all users			
- Also consider creating the upload button a button on it’s own rather than being in a field. 		
- The screen reader should recognise the button is specifically going to upload images from the user’s device.

![Screenshot 2019-06-24 at 17 04 14](https://user-images.githubusercontent.com/877585/60034281-2b7d2180-96a2-11e9-9da3-23eb5d9c2209.png)

**Comment by zerolab** [comment]
> "Upload images" – Consider renaming the field name to something more meaningful.

Do consider #4055 for this. That is possibly a good one to tackle in the accessibility improvements work
